### PR TITLE
Tokenizer APIs Update

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/machinelearning/issues/7158
Fixes https://github.com/dotnet/machinelearning/issues/7144

- https://github.com/dotnet/machinelearning/issues/7158
- https://github.com/dotnet/machinelearning/issues/7144

This change addresses all feedback we got in the design review.

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @michaelgsharp @luisquintanilla @ericstj @stephentoub 

@LittleLittleCloud the change here is adding the support to the Added tokens in sentence piece Bpe tokenizers.

File: test/Microsoft.ML.Tokenizers.Tests/LlamaTests.cs [hunk 1]
```diff
@@ -19,20 +20,44 @@ public class LlamaTests
...
 
         private static Tokenizer CreateLMistralTokenizer()
         {
             // @"https://huggingface.co/mistralai/Mistral-7B-v0.1/resolve/main/tokenizer.model?download=true";
             using Stream remoteStream = File.OpenRead(Path.Combine(@"Mistral", "tokenizer.model"));
-            return Tokenizer.CreateLlama(remoteStream);
+            return LlamaTokenizer.Create(remoteStream);
+        }
+
+        private static Tokenizer CreateLPhi3Tokenizer()
```
Comment (LittleLittleCloud) [inline test/Microsoft.ML.Tokenizers.Tests/LlamaTests.cs:41]: `CreatePhi3Tokenizer`?

File: test/Microsoft.ML.Tokenizers.Tests/LlamaTests.cs [hunk 2]
```diff
@@ -511,19 +578,74 @@ public void TestTokenizerCountTokens(string text, string normalizedText, string[
...
             Assert.Equal(normalizedText, normalizedString);
             Assert.Equal(7, tokenCount);
-            Assert.Equal(expectedOffsets[expectedOffsets.Length - 7].Index, tokenizer.LastIndexOfTokenCount(text.AsSpan(), 7, out normalizedString, out tokenCount));
+            Assert.Equal(expectedOffsets[expectedOffsets.Length - 7].Index, tokenizer.GetIndexByTokenCountFromEnd(text.AsSpan(), 7, out normalizedString, out tokenCount));
             Assert.Equal(normalizedText, normalizedString);
             Assert.Equal(7, tokenCount);
         }
+
+        [Fact]
+        public void TestPhi3Tokenizer()
```
Comment (LittleLittleCloud) [inline test/Microsoft.ML.Tokenizers.Tests/LlamaTests.cs:597]: Can you also cover Phi3Tokenizer with the following cases

```csharp
var messages = new string[]
        {
            "Can you provide ways to eat combinations of bananas and dragonfruits?",
            "Sure! Here are some ways to eat bananas and dragonfruits together: 1. Banana and dragonfruit smoothie: Blend bananas and dragonfruits together with some milk and honey. 2. Banana and dragonfruit salad: Mix sliced bananas and dragonfruits together with some lemon juice and honey.",
            "What about solving an 2x + 3 = 7 equation?",
            "\nCount to 3\n",
            "<|user|>",
            "<|end|>",
            "<|assistant|>",
            "<|user|>\nCount to 3<|end|>\n<|assistant|>",
        };
```

And corresponding encoding result
```result
1, 1815, 366, 3867, 5837, 304, 17545, 18240, 310, 9892, 16397, 322, 8338, 265, 29888, 21211, 29973
1, 18585, 29991, 2266, 526, 777, 5837, 304, 17545, 9892, 16397, 322, 8338, 265, 29888, 21211, 4208, 29901, 29871, 29896, 29889, 10765, 1648, 322, 8338, 265, 29888, 9216, 10597, 347, 29901, 3164, 355, 9892, 16397, 322, 8338, 265, 29888, 21211, 4208, 411, 777, 27274, 322, 298, 4992, 29889, 29871, 29906, 29889, 10765, 1648, 322, 8338, 265, 29888, 9216, 4497, 328, 29901, 23478, 269, 506, 287, 9892, 16397, 322, 8338, 265, 29888, 21211, 4208, 411, 777, 454, 3712, 3623, 625, 322, 298, 4992, 29889
1, 1724, 1048, 17069, 385, 29871, 29906, 29916, 718, 29871, 29941, 353, 29871, 29955, 6306, 29973
1, 29871, 13, 3981, 304, 29871, 29941, 13
1, 32010
1, 32007
1, 32001
1, 32010, 29871, 13, 3981, 304, 29871, 29941, 32007, 29871, 13, 32001
```

**Review by michaelgsharp — APPROVED** [review]
LGTM.

**Pull request merged**